dpnp/dpnp_iface_mathematical.py

```diff
@@ -854,42 +855,109 @@ def cumprod(x1, **kwargs):
+        size as `a`, and the same shape as `a` if `axis` is not ``None`` or `a`
+        is a 1-d array.
 
     See Also
     --------
-    :obj:`dpnp.diff` : Calculate the n-th discrete difference along the given axis.
+    :obj:`dpnp.sum` : Sum array elements.
+    :obj:`dpnp.diff` : Calculate the n-th discrete difference along given axis.
 
     Examples
     --------
     >>> import dpnp as np
-    >>> a = np.array([1, 2, 4])
-    >>> result = np.cumsum(a)
-    >>> [x for x in result]
-    [1, 2, 7]
-    >>> b = np.array([[1, 2, 3], [4, 5, 6]])
-    >>> result = np.cumsum(b)
-    >>> [x for x in result]
-    [1, 2, 6, 10, 15, 21]
+    >>> a = np.array([[1, 2, 3], [4, 5, 6]])
+    >>> a
+    array([[1, 2, 3],
+           [4, 5, 6]])
+    >>> np.cumsum(a)
+    array([ 1,  3,  6, 10, 15, 21])
+    >>> np.cumsum(a, dtype=float)     # specifies type of output value(s)
+    array([ 1.,  3.,  6., 10., 15., 21.])
+
+    >>> np.cumsum(a, axis=0)     # sum over rows for each of the 3 columns
+    array([[1, 2, 3],
+           [5, 7, 9]])
+    >>> np.cumsum(a, axis=1)     # sum over columns for each of the 2 rows
+    array([[ 1,  3,  6],
+           [ 4,  9, 15]])
+
+    ``cumsum(b)[-1]`` may not be equal to ``sum(b)``
+
+    >>> b = np.array([1, 2e-9, 3e-9] * 10000)
+    >>> b.cumsum().dtype == b.sum().dtype == np.float64
+    True
+    >>> b.cumsum()[-1] == b.sum()
+    array(False)
 
     """
 
-    x1_desc = dpnp.get_dpnp_descriptor(x1, copy_when_nondefault_queue=False)
-    if x1_desc and not kwargs:
-        return dpnp_cumsum(x1_desc).get_pyobj()
+    dpnp.check_supported_arrays_type(a)
+    if a.ndim > 1 and axis is None:
+        usm_a = dpnp.ravel(a).get_array()
+    else:
+        usm_a = dpnp.get_usm_ndarray(a)
+
+    input_out = out
+    if out is None:
+        usm_out = None
+    else:
+        dpnp.check_supported_arrays_type(out)
+
+        # get dtype used by dpctl for result array in cumulative_sum
+        if dtype is None:
+            res_dt = dtu._default_accumulation_dtype(a.dtype, a.sycl_queue)
+        else:
+            res_dt = dpnp.dtype(dtype)
+            res_dt = dtu._to_device_supported_dtype(res_dt, a.sycl_device)
+
+        # dpctl requires strict data type matching of out array with the result
+        if out.dtype != res_dt:
+            out = dpnp.astype(out, dtype=res_dt, copy=False)
+
+        usm_out = dpnp.get_usm_ndarray(out)
 
-    return call_origin(numpy.cumsum, x1, **kwargs)
+    res_usm = dpt.cumulative_sum(usm_a, axis=axis, dtype=dtype, out=usm_out)
+    if out is None:
+        return dpnp_array._create_from_usm_ndarray(res_usm)
+
+    if input_out is not out:
+        # need to copy back from temporary memory to input `out`
+        dpnp.copyto(input_out, out, casting="unsafe")
+
+    if not isinstance(input_out, dpnp_array):
+        return dpnp_array._create_from_usm_ndarray(input_out)
+    else:
+        return input_out
```

Comment:
With the current implementation, the following is False
```
import dpctl,dpctl.tensor as dpt, numpy, dpnp
a=dpnp.ones((5,5))
out=dpt.empty((5, 5))
res=dpnp.cumsum(a, out=out, axis=1)
res is out
# False
```
I guess we do not need to change usm_ndarray to dpnp_array
```suggestion
   return input_out
```

Reply:
We have quite the same logic for `dpnp.sum`:
```python
a=dpnp.ones((5,5))
out=dpt.empty((5, ))
res=dpnp.sum(a, out=out, axis=1)

res is out
# Out: False
```
So I will update the docstring to reflect actual behavior there.